### PR TITLE
Workflow destroy service

### LINKED ISSUE #249: Workflow: Deletion Service
## Summary
After a `WorfklowExecution` has either completed or errored, a user may want to delete it. Since this doesn't require a call to the WES API we can just delete immediately after the user clicks the `Delete` button for a workflow

## Acceptance Criteria
* Create a `WorkflowExecutions::DeletionService`
  * Accept a `workflow_execution` as parameters
  * Validate that the `workflow_execution` is in one of the `deletable` states (either completed, or errored)
  * If it is in one of the above states, then the workflow and it's children should be destroyed
  * A controller under workflow_executions
  * UI updated to include `Delete` on the workflow executions table listing

### PR DESCRIPTION
**Pull request by ksierks**
## What does this PR do and why?
A user can delete a WorkflowExecution after it has been completed, errored, or canceled.
Fixes #249.

## Screenshots or screen recordings
Screenshot from 2024-03-05 08-55-47
Screenshot from 2024-03-05 08-55-56
image

## How to set up and validate locally
1. Login to irida-next.
2. Create a new workflow execution.
3. Open a rails console and run the following:
```
w = WorkflowExecution.last
w.state = "completed"
w.save
```
5. Refresh the workflow execution page.
6. Click the `Delete` link.
7. Confirm the deletion.
8. Verify the workflow execution has been deleted.

Repeat with `w.state = "error"` and `w.state = "canceled"`.

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: app/services/workflow_executions/destroy_service.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+module WorkflowExecutions
+  # Service used to delete a WorkflowExecution
+  class DestroyService < BaseService
+    def initialize(workflow_execution, user = nil, params = {})
+      super(user, params)
+      @workflow_execution = workflow_execution
+    end
+
+    def execute
+      return unless @workflow_execution.completed? || @workflow_execution.error?
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/services/workflow_executions/destroy_service.rb]: dbe3b1ad92f99f4c34e3399097b12d2402ce75d9

File: app/views/workflow_executions/_workflow_execution.html.erb [hunk 2]
```diff
@@ -1,32 +1,56 @@
...
+        <%= form.submit t(:"workflow_executions.index.actions.cancel_button"),
+                    class:
+                      "font-medium text-blue-600 underline dark:text-blue-500 hover:no-underline cursor-pointer",
+                    data: {
+                      turbo_confirm:
+                        t(:"workflow_executions.index.actions.cancel_confirm")
+                    } %>
+      <% end %>
+    <% end %>
+    <% if workflow_execution.completed? || workflow_execution.error? %>
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/views/workflow_executions/_workflow_execution.html.erb]: dbe3b1ad92f99f4c34e3399097b12d2402ce75d9

File: app/services/workflow_executions/destroy_service.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+      @workflow_execution.destroy
```
[the remark this answers is not in the document]
Comment (ksierks) [reply app/services/workflow_executions/destroy_service.rb:14]: Is this okay? 1b70723134da37196e79e1f9d81eb9bf10443942.

[the remark this answers is not in the document]
Comment (ericenns) [reply app/services/workflow_executions/destroy_service.rb:14]: Yup that looks good

File: test/controllers/workflow_executions_controller_test.rb [hunk 4]
```diff
@@ -58,4 +58,54 @@ class WorfklowExecutionsControllerTest < ActionDispatch::IntegrationTest
...
+
+  test 'should delete a canceled workflow' do
+    workflow_execution = workflow_executions(:irida_next_example_canceled)
+    assert workflow_execution.canceled?
+    assert_difference -> { WorkflowExecution.count } => -1,
+                      -> { SamplesWorkflowExecution.count } => -1 do
+      delete workflow_execution_path(workflow_execution, format: :turbo_stream)
+    end
+    assert_response :success
+  end
```
Comment (deepsidhu85) [inline test/controllers/workflow_executions_controller_test.rb:110]: Can I get you to add tests for `submitting`, `new`, `queued`, and `running` as well

File: test/services/workflow_executions/destroy_service_test.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      assert workflow_execution.canceled?
+
+      assert_difference -> { WorkflowExecution.count } => -1,
+                        -> { SamplesWorkflowExecution.count } => -1,
+                        -> { Sample.count } => 0 do
+        WorkflowExecutions::DestroyService.new(workflow_execution, @user).execute
+      end
+    end
+  end
+end
```
Comment (deepsidhu85) [inline test/services/workflow_executions/destroy_service_test.rb:75]: Same comment as above for the controller test

File: test/system/workflow_executions_test.rb [hunk 6]
```diff
@@ -35,4 +37,97 @@ class WorkflowExecutionsTest < ApplicationSystemTestCase
...
+      assert_selector 'td:nth-child(4)', text: workflow_execution.state
+      assert_selector 'input[type="submit"][value="Delete"]', count: 1
+      find('input[type="submit"][value="Delete"]').click
+    end
+
+    assert_text 'Confirmation required'
+    click_button 'Confirm'
+
+    assert_no_text workflow_execution.run_id
+  end
```
Comment (deepsidhu85) [inline test/system/workflow_executions_test.rb:132]: Same as above

**Review by deepsidhu85 — CHANGES_REQUESTED** [review]
This looks great, just have a couple of comments to add a few more tests

File: test/system/workflow_executions_test.rb (hunk 6 above)
Comment (ksierks) [reply test/system/workflow_executions_test.rb:147]: 2e4c95fa7d54e38f495de3a628c534eb7d1a7a42

File: test/services/workflow_executions/destroy_service_test.rb (hunk 5 above)
Comment (ksierks) [reply test/services/workflow_executions/destroy_service_test.rb:115]: 2e4c95fa7d54e38f495de3a628c534eb7d1a7a42

File: test/controllers/workflow_executions_controller_test.rb (hunk 4 above)
Comment (ksierks) [reply test/controllers/workflow_executions_controller_test.rb:120]: 2e4c95fa7d54e38f495de3a628c534eb7d1a7a42

Comment (deepsidhu85) [reply test/controllers/workflow_executions_controller_test.rb:120]: I tried creating my own state :laughing: Thanks!

**Review by ericenns — APPROVED** [review]

**Review by deepsidhu85 — APPROVED** [review]
Looks good to me!

**Pull request merged**